src/java/org/apache/cassandra/db/compression/ZstdCompressionDictionary.java

```diff
@@ -0,0 +1,217 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.compression;
+
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.luben.zstd.ZstdDictCompress;
+import com.github.luben.zstd.ZstdDictDecompress;
+import org.apache.cassandra.io.compress.ZstdCompressorBase;
+import org.apache.cassandra.utils.concurrent.Ref;
+import org.apache.cassandra.utils.concurrent.RefCounted;
+import org.apache.cassandra.utils.concurrent.SelfRefCounted;
+
+public class ZstdCompressionDictionary implements CompressionDictionary, SelfRefCounted<ZstdCompressionDictionary>
+{
+    private static final Logger logger = LoggerFactory.getLogger(ZstdCompressionDictionary.class);
+
+    private final DictId dictId;
+    private final byte[] rawDictionary;
+    // One ZstdDictDecompress and multiple ZstdDictCompress (per level) can be derived from the same raw dictionary content
+    private final ConcurrentHashMap<Integer, ZstdDictCompress> zstdDictCompressPerLevel = new ConcurrentHashMap<>();
+    private volatile ZstdDictDecompress dictDecompress;
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final Ref<ZstdCompressionDictionary> selfRef;
+
+    public ZstdCompressionDictionary(DictId dictId, byte[] rawDictionary)
+    {
+        this.dictId = dictId;
+        this.rawDictionary = rawDictionary;
+        this.selfRef = new Ref<>(this, new Tidy(zstdDictCompressPerLevel, dictDecompress));
+    }
+
+    @Override
+    public DictId identifier()
+    {
+        return dictId;
+    }
+
+    @Override
+    public Kind kind()
+    {
+        return Kind.ZSTD;
+    }
+
+    @Override
+    public byte[] rawDictionary()
+    {
+        return rawDictionary;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (!(o instanceof ZstdCompressionDictionary)) return false;
+        ZstdCompressionDictionary that = (ZstdCompressionDictionary) o;
+        return Objects.equals(dictId, that.dictId);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return dictId.hashCode();
+    }
+
+    /**
+     * Get a pre-processed compression tables that is optimized for compression.
+     * It is derived/computed from dictionary bytes.
+     * The internal data structure is different from the tables for decompression.
+     *
+     * @param compressionLevel compression level to create the compression table
+     * @return ZstdDictCompress
+     */
+    public ZstdDictCompress dictionaryForCompression(int compressionLevel)
+    {
+        if (closed.get())
+            throw new IllegalStateException("Dictionary has been closed");
```

Comment:
maybe being more descriptive in this message (and all other same messages) by including dict id with kind and id would be nice.